integration-tests/tests/src/tests/sync/flexible.ts

```diff
@@ -0,0 +1,1361 @@
+          const subs = this.realm.subscriptions;
+          expect(subs.error).to.be.null;
+        });
+
+        it("is null if there was no error synchronising subscriptions", async function (this: RealmContext) {
+          const { subs } = addSubscriptionForPerson(this.realm);
+          await subs.waitForSynchronization();
+
+          expect(subs.error).to.be.null;
+        });
+
+        it("contains the error message if there was an error synchronising subscriptions", async function (this: RealmContext) {
+          const { subs } = addSubscription(
+            this.realm,
+            this.realm.objects(FlexiblePersonSchema.name).filtered("nonQueryable == 'test'"),
+          );
+          await expect(subs.waitForSynchronization()).to.be.rejected;
+
+          expect(subs.error).to.equal(
+            'Client provided query with bad syntax: invalid match expression for table "Person": key "nonQueryable" is not a queryable field',
+          );
+        });
+
+        it("is null if there was an error but it was subsequently corrected", async function (this: RealmContext) {
+          const { subs } = addSubscription(
+            this.realm,
+            this.realm.objects(FlexiblePersonSchema.name).filtered("nonQueryable == 'test'"),
+          );
+          await expect(subs.waitForSynchronization()).to.be.rejected;
+
+          expect(subs.error).to.equal(
+            'Client provided query with bad syntax: invalid match expression for table "Person": key "nonQueryable" is not a queryable field',
+          );
+
+          subs.update((mutableSubs) => {
+            mutableSubs.removeAll();
+            mutableSubs.add(this.realm.objects(FlexiblePersonSchema.name));
+          });
+          await expect(subs.waitForSynchronization()).to.be.fulfilled;
+
+          expect(subs.error).to.be.null;
+        });
+
+        it("still contains the erroring subscription in the set if there was an error synchronising", async function (this: RealmContext) {
+          const { subs } = addSubscription(
+            this.realm,
+            this.realm.objects(FlexiblePersonSchema.name).filtered("nonQueryable == 'test'"),
+          );
+          await expect(subs.waitForSynchronization()).to.be.rejected;
+
+          expect(subs).to.have.length(1);
+        });
+      });
+
+      describe("#waitForSynchronization", async function () {
+        it("returns a promise", function (this: RealmContext) {
+          const { subs } = addSubscriptionForPerson(this.realm);
+
+          const promise = subs.waitForSynchronization();
+          expect(promise).to.be.instanceOf(Promise);
+          return promise;
+        });
+
+        it("throws an error if no subscriptions have been created", async function (this: RealmContext) {
+          expect(this.realm.subscriptions.waitForSynchronization()).to.be.rejectedWith(
+            "`waitForSynchronisation` cannot be called before creating a subscription set using `update`",
+          );
+        });
```

Comment:
Yeah it's an interesting point. I'll bring this up for discussion and see if we can unify on an approach to it. I think if this is the desired behaviour, ideally core would behave like that rather than having SDK specific logic